### ISSUE TITLE
[Flaky Test] [k8s.io] MirrorPodWithGracePeriod when create a mirror pod mirror pod termination should satisfy grace period when static pod is deleted [NodeConformance]

### ISSUE THREAD

**Issue by hasheddan** [kubernetes/kubernetes#94223]
**Which jobs are flaking**:

`node-kubelet-master (ci-kubernetes-node-kubelet)`

**Which test(s) are flaking**:

`[k8s.io] MirrorPodWithGracePeriod when create a mirror pod mirror pod termination should satisfy grace period when static pod is deleted [NodeConformance]`

**Testgrid link**:

https://testgrid.k8s.io/sig-release-master-blocking#node-kubelet-master

**Reason for failure**:

```
_output/local/go/src/k8s.io/kubernetes/test/e2e_node/mirror_pod_grace_period_test.go:54
Unexpected error:
    <*errors.StatusError | 0xc000f2ab40>: {
        ErrStatus: {
            TypeMeta: {Kind: "", APIVersion: ""},
            ListMeta: {
                SelfLink: "",
                ResourceVersion: "",
                Continue: "",
                RemainingItemCount: nil,
            },
            Status: "Failure",
            Message: "pods \"graceful-pod-318b04ef-32f9-4be7-be28-72dbb4dcc6e5-tmp-node-e2e-0ead0b6b-cos-81-12871-1190-0\" not found",
            Reason: "NotFound",
            Details: {
                Name: "graceful-pod-318b04ef-32f9-4be7-be28-72dbb4dcc6e5-tmp-node-e2e-0ead0b6b-cos-81-12871-1190-0",
                Group: "",
                Kind: "pods",
                UID: "",
                Causes: nil,
                RetryAfterSeconds: 0,
            },
            Code: 404,
        },
    }
    pods "graceful-pod-318b04ef-32f9-4be7-be28-72dbb4dcc6e5-tmp-node-e2e-0ead0b6b-cos-81-12871-1190-0" not found
occurred
_output/local/go/src/k8s.io/kubernetes/test/e2e_node/mirror_pod_grace_period_test.go:72
```

**Anything else we need to know**:

/cc @kubernetes/ci-signal @kubernetes/sig-node-bugs 
/sig node
/priority critical-urgent
/milestone v1.19

**Comment by sjenning** [comment]
History of this failure

https://storage.googleapis.com/k8s-gubernator/triage/index.html?pr=1&test=MirrorPodWithGracePeriod

Looking at the code for this test, it isn't a flake as much as a continued bug in the kubelet with removing mirror pods too early.  In other words, the test is catching the case for which it is designed.

**Comment by dims** [comment]
@sjenning sounds like NOT a blocker for 1.19.0 :)

**Comment by sjenning** [comment]
Yes @dims, not a blocker.

Looking at the logs
```
Aug 23 02:42:06 tmp-node-e2e-03dde4d6-ubuntu-gke-1804-1-15-v20200821 kubelet[4123]: I0823 02:42:06.917163    4123 kubelet.go:1837] SyncLoop (REMOVE, "file"): "graceful-pod-b7adf57b-c69f-4b50-a296-b5a420e8489c-tmp-node-e2e-03dde4d6-ubuntu-gke-1804-1-15-v20200821_mirror-pod-with-grace-period-5533(7cc88844d2e7f377323d7c9f9d8c9157)"
Aug 23 02:42:06 tmp-node-e2e-03dde4d6-ubuntu-gke-1804-1-15-v20200821 kubelet[4123]: I0823 02:42:06.917237    4123 kubelet_pods.go:1247] Killing unwanted pod "graceful-pod-b7adf57b-c69f-4b50-a296-b5a420e8489c-tmp-node-e2e-03dde4d6-ubuntu-gke-1804-1-15-v20200821"
Aug 23 02:42:06 tmp-node-e2e-03dde4d6-ubuntu-gke-1804-1-15-v20200821 kubelet[4123]: I0823 02:42:06.930412    4123 plugins.go:422] Calling network plugin kubenet to tear down pod "graceful-pod-b7adf57b-c69f-4b50-a296-b5a420e8489c-tmp-node-e2e-03dde4d6-ubuntu-gke-1804-1-15-v20200821_mirror-pod-with-grace-period-5533"
Aug 23 02:42:06 tmp-node-e2e-03dde4d6-ubuntu-gke-1804-1-15-v20200821 kubelet[4123]: I0823 02:42:06.931761    4123 kubenet_linux.go:740] Removing mirror-pod-with-grace-period-5533/graceful-pod-b7adf57b-c69f-4b50-a296-b5a420e8489c-tmp-node-e2e-03dde4d6-ubuntu-gke-1804-1-15-v20200821 from 'kubenet-loopback' with CNI 'loopback' plugin and runtime: &{ContainerID:622bb2a2f2a9239d47f5bfd53c9522ec8f2e03f352e7b4ee2313d5fdf6bec354 NetNS:/proc/8715/ns/net IfName:lo Args:[] CapabilityArgs:map[] CacheDir:/tmp/node-e2e-20200823T022854/cni/cache}
Aug 23 02:42:06 tmp-node-e2e-03dde4d6-ubuntu-gke-1804-1-15-v20200821 kubelet[4123]: I0823 02:42:06.936149    4123 kubenet_linux.go:740] Removing mirror-pod-with-grace-period-5533/graceful-pod-b7adf57b-c69f-4b50-a296-b5a420e8489c-tmp-node-e2e-03dde4d6-ubuntu-gke-1804-1-15-v20200821 from 'kubenet' with CNI 'bridge' plugin and runtime: &{ContainerID:622bb2a2f2a9239d47f5bfd53c9522ec8f2e03f352e7b4ee2313d5fdf6bec354 NetNS:/proc/8715/ns/net IfName:eth0 Args:[] CapabilityArgs:map[] CacheDir:/tmp/node-e2e-20200823T022854/cni/cache}
Aug 23 02:42:06 tmp-node-e2e-03dde4d6-ubuntu-gke-1804-1-15-v20200821 kubelet[4123]: I0823 02:42:06.978743    4123 kubenet_linux.go:582] TearDownPod took 48.303778ms for mirror-pod-with-grace-period-5533/graceful-pod-b7adf57b-c69f-4b50-a296-b5a420e8489c-tmp-node-e2e-03dde4d6-ubuntu-gke-1804-1-15-v20200821
Aug 23 02:42:06 tmp-node-e2e-03dde4d6-ubuntu-gke-1804-1-15-v20200821 kubelet[4123]: I0823 02:42:06.986196    4123 kuberuntime_manager.go:933] getSandboxIDByPodUID got sandbox IDs ["622bb2a2f2a9239d47f5bfd53c9522ec8f2e03f352e7b4ee2313d5fdf6bec354"] for pod "graceful-pod-b7adf57b-c69f-4b50-a296-b5a420e8489c-tmp-node-e2e-03dde4d6-ubuntu-gke-1804-1-15-v20200821_mirror-pod-with-grace-period-5533(7cc88844d2e7f377323d7c9f9d8c9157)"
Aug 23 02:42:07 tmp-node-e2e-03dde4d6-ubuntu-gke-1804-1-15-v20200821 kubelet[4123]: I0823 02:42:07.051713    4123 generic.go:386] PLEG: Write status for graceful-pod-b7adf57b-c69f-4b50-a296-b5a420e8489c-tmp-node-e2e-03dde4d6-ubuntu-gke-1804-1-15-v20200821/mirror-pod-with-grace-period-5533: &container.PodStatus{ID:"7cc88844d2e7f377323d7c9f9d8c9157", Name:"graceful-pod-b7adf57b-c69f-4b50-a296-b5a420e8489c-tmp-node-e2e-03dde4d6-ubuntu-gke-1804-1-15-v20200821", Namespace:"mirror-pod-with-grace-period-5533", IPs:[]string{}, ContainerStatuses:[]*container.Status{(*container.Status)(0xc001504c40)}, SandboxStatuses:[]*v1alpha2.PodSandboxStatus{(*v1alpha2.PodSandboxStatus)(0xc001913260)}} (err: <nil>)
Aug 23 02:42:07 tmp-node-e2e-03dde4d6-ubuntu-gke-1804-1-15-v20200821 kubelet[4123]: I0823 02:42:07.266889    4123 manager.go:1044] Destroyed container: "/kubepods/besteffort/pod7cc88844d2e7f377323d7c9f9d8c9157/622bb2a2f2a9239d47f5bfd53c9522ec8f2e03f352e7b4ee2313d5fdf6bec354" (aliases: [k8s_POD_graceful-pod-b7adf57b-c69f-4b50-a296-b5a420e8489c-tmp-node-e2e-03dde4d6-ubuntu-gke-1804-1-15-v20200821_mirror-pod-with-grace-period-5533_7cc88844d2e7f377323d7c9f9d8c9157_0 622bb2a2f2a9239d47f5bfd53c9522ec8f2e03f352e7b4ee2313d5fdf6bec354], namespace: "docker")
Aug 23 02:42:07 tmp-node-e2e-03dde4d6-ubuntu-gke-1804-1-15-v20200821 kubelet[4123]: I0823 02:42:07.279029    4123 kubelet_pods.go:1247] Killing unwanted pod "graceful-pod-b7adf57b-c69f-4b50-a296-b5a420e8489c-tmp-node-e2e-03dde4d6-ubuntu-gke-1804-1-15-v20200821"
Aug 23 02:42:07 tmp-node-e2e-03dde4d6-ubuntu-gke-1804-1-15-v20200821 kubelet[4123]: I0823 02:42:07.290535    4123 mirror_client.go:125] Deleting a mirror pod "graceful-pod-b7adf57b-c69f-4b50-a296-b5a420e8489c-tmp-node-e2e-03dde4d6-ubuntu-gke-1804-1-15-v20200821_mirror-pod-with-grace-period-5533" (uid (*types.UID)(nil))
Aug 23 02:42:07 tmp-node-e2e-03dde4d6-ubuntu-gke-1804-1-15-v20200821 kubelet[4123]: I0823 02:42:07.297518    4123 kubelet_pods.go:1044] deleted pod "graceful-pod-b7adf57b-c69f-4b50-a296-b5a420e8489c-tmp-node-e2e-03dde4d6-ubuntu-gke-1804-1-15-v20200821_mirror-pod-with-grace-period-5533"
Aug 23 02:42:07 tmp-node-e2e-03dde4d6-ubuntu-gke-1804-1-15-v20200821 kubelet[4123]: I0823 02:42:07.297684    4123 kubelet.go:1843] SyncLoop (DELETE, "api"): "graceful-pod-b7adf57b-c69f-4b50-a296-b5a420e8489c-tmp-node-e2e-03dde4d6-ubuntu-gke-1804-1-15-v20200821_mirror-pod-with-grace-period-5533(65e0d673-3c30-4e79-882f-0eec302eb2bd)"
```
Deletion of the mirror pod occurs <1s after the file-based DELETE comes in.

**Comment by hasheddan** [comment]
I am going to go ahead and bump to v1.20 then :)

/milestone v1.20

**Comment by sjenning** [comment]
also see previous bug caught by this test https://github.com/kubernetes/kubernetes/issues/93210
fyi @liggitt

**Comment by derekwaynecarr** [comment]
i feel like we should add more depth to how we check if the mirror pod containers are running beyond looking at pod phase.

will open some additional prs to help get more debug data.

**Comment by derekwaynecarr** [comment]
the more i look at the log, the more i feel like there is a bug this exposes for when we report phase=running.

**Comment by hasheddan** [comment]
@derekwaynecarr would you consider this bug release blocking?

**Comment by derekwaynecarr** [comment]
@hasheddan no, i think the more likely outcome is that the test is looking at pod phase running which is not always a good thing to look at because a pod can be running even if some of its containers are in an unknown state.  definitely not a blocker.

**Comment by harche** [comment]
Maybe similar to, https://github.com/kubernetes/kubernetes/issues/97722

**Comment by ehashman** [comment]
Can this be closed now? I think this may have been fixed when #97722 was fixed.
